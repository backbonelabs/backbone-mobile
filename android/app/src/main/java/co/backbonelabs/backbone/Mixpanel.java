package co.backbonelabs.backbone;

import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.bridge.ReadableType;
import com.mixpanel.android.mpmetrics.MixpanelAPI;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * React Native wrapper for the Mixpanel Android API
 * Most of this code was copied from https://github.com/davodesign84/react-native-mixpanel
 */
public class Mixpanel extends ReactContextBaseJavaModule implements LifecycleEventListener {
    private MixpanelAPI mixpanel;

    public Mixpanel(ReactApplicationContext reactContext) {
        super(reactContext);
        String projectToken = BuildConfig.MIXPANEL_TOKEN;
        mixpanel = MixpanelAPI.getInstance(reactContext.getCurrentActivity(), projectToken);
    }

    /**
     * Converts a ReadableMap object from React Native to a JSONObject.
     * @param readableMap The ReadableMap object to convert to a JSONObject
     * @return JSONObject equivalent of readableMap
     * @throws JSONException
     */
    static JSONObject reactToJSON(ReadableMap readableMap) throws JSONException {
        JSONObject jsonObject = new JSONObject();
        ReadableMapKeySetIterator iterator = readableMap.keySetIterator();
        while(iterator.hasNextKey()){
            String key = iterator.nextKey();
            ReadableType valueType = readableMap.getType(key);
            switch (valueType){
                case Null:
                    jsonObject.put(key,JSONObject.NULL);
                    break;
                case Boolean:
                    jsonObject.put(key, readableMap.getBoolean(key));
                    break;
                case Number:
                    jsonObject.put(key, readableMap.getDouble(key));
                    break;
                case String:
                    jsonObject.put(key, readableMap.getString(key));
                    break;
                case Map:
                    jsonObject.put(key, reactToJSON(readableMap.getMap(key)));
                    break;
                case Array:
                    jsonObject.put(key, reactToJSON(readableMap.getArray(key)));
                    break;
            }
        }

        return jsonObject;
    }

    /**
     * Converts a ReadableArray object from React Native to a JSONObject.
     * @param readableArray The ReadableArray object to convert to a JSONObject
     * @return JSONObject equivalent of readableArray
     * @throws JSONException
     */
    static JSONArray reactToJSON(ReadableArray readableArray) throws JSONException {
        JSONArray jsonArray = new JSONArray();
        for(int i=0; i < readableArray.size(); i++) {
            ReadableType valueType = readableArray.getType(i);
            switch (valueType){
                case Null:
                    jsonArray.put(JSONObject.NULL);
                    break;
                case Boolean:
                    jsonArray.put(readableArray.getBoolean(i));
                    break;
                case Number:
                    jsonArray.put(readableArray.getDouble(i));
                    break;
                case String:
                    jsonArray.put(readableArray.getString(i));
                    break;
                case Map:
                    jsonArray.put(reactToJSON(readableArray.getMap(i)));
                    break;
                case Array:
                    jsonArray.put(reactToJSON(readableArray.getArray(i)));
                    break;
            }
        }
        return jsonArray;
    }

    @Override
    public String getName() {
        return "Mixpanel";
    }

    /**
     * Track an event with no properties
     * @param name The name of the event to send
     */
    @ReactMethod
    public void track(final String name) {
        mixpanel.track(name);
    }

    /**
     * Track an event with properties
     * @param name The name of the event to send
     * @param props Key-value pairs of the properties to include in the event
     */
    @ReactMethod
    public void trackWithProperties(final String name, final ReadableMap props) {
        JSONObject obj = null;
        try {
            obj = Mixpanel.reactToJSON(props);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        mixpanel.track(name, obj);
    }

    /**
     * Create an alias for linking a user ID to the current ID
     * @param newId New ID to link to current ID
     */
    @ReactMethod
    public void createAlias(final String newId) {
        mixpanel.alias(newId, mixpanel.getDistinctId());
    }

    /**
     * Associate future calls to track(String) with the user identified by the given distinct ID
     * @param userId A unique identifier for the user
     */
    @ReactMethod
    public void identify(final String userId) {
        mixpanel.identify(userId);
        mixpanel.getPeople().identify(userId);
    }

    /**
     * Begin timing of an event
     * @param event The name of the event to track with timing
     */
    @ReactMethod
    public void timeEvent(final String event) {
        mixpanel.timeEvent(event);
    }

    /**
     * Register properties that will be sent with every subsequent call to track(String)
     * @param properties Super properties to register
     */
    @ReactMethod
    public void registerSuperProperties(final ReadableMap properties) {
        JSONObject obj = null;
        try {
            obj = Mixpanel.reactToJSON(properties);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        mixpanel.registerSuperProperties(obj);
    }

    /**
     * Register super properties like registerSuperProperties(ReadableMap), but only
     * if no other super property with the same names have already been registered
     * @param properties Super properties to register
     */
    @ReactMethod
    public void registerSuperPropertiesOnce(final ReadableMap properties) {
        JSONObject obj = null;
        try {
            obj = Mixpanel.reactToJSON(properties);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        mixpanel.registerSuperPropertiesOnce(obj);
    }

    /**
     * Set a collection of properties on a user all at once
     * @param properties Key-value pair of property names and values
     */
    @ReactMethod
    public void set(final ReadableMap properties) {
        JSONObject obj = null;
        try {
            obj = Mixpanel.reactToJSON(properties);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        mixpanel.getPeople().set(obj);
    }

    /**
     * Increment an existing numbered user property
     * @param property The user property to update
     * @param count The amount to be added to the current value of the property
     */
    @ReactMethod
    public void incrementUserPropertyBy(final String property, final double count) {
        mixpanel.getPeople().increment(property, count);
    }

    /**
     * Clear all distinct IDs, super properties, and push registrations from persistent storage,
     * and then flush the events
     */
    @ReactMethod
    public void reset() {
        mixpanel.reset();
        mixpanel.flush();
    }

    /**
     * Push all queued events and user profile changes to Mixpanel
     */
    @ReactMethod
    public void flush() {
        mixpanel.flush();
    }

    @Override
    public void onHostResume() {
        // Actvity `onResume`
    }

    @Override
    public void onHostPause() {
        // Actvity `onPause`

        if (mixpanel != null) {
            mixpanel.flush();
        }
    }

    @Override
    public void onHostDestroy() {
        // Actvity `onDestroy`

        if (mixpanel != null) {
            mixpanel.flush();
        }
    }
}
