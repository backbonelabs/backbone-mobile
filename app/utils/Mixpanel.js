import { NativeModules } from 'react-native';
import { map } from 'lodash';
import constants from './constants';

const { Mixpanel, BluetoothService } = NativeModules;

export default {
  /**
   * Associates all future events with the specified user id
   * @param  {String}  userId  User account id
   */
  identify(userId) {
    Mixpanel.identify(userId);
  },

  /**
   * Sets the user's profile properties on Mixpanel
   * @param  {Object}  userDetails  User profile properties
   */
  set(userDetails) {
    Mixpanel.set(userDetails);
  },

  /**
   * Formats the user profile properties before setting on Mixpanel
   * @param  {Object}  userDetails  User profile properties
   */
  setUserProperties(user) {
    const userClone = Object.assign({ ...user }, { ...user.settings });
    const { gender, heightUnitPreference, weightUnitPreference } = user;
    const { height, weight } = constants;
    const formattedGender = gender === constants.gender.male ? 'male' : 'female';

    // Delete unwanted profile properties
    [
      '_id',
      'confirmationToken',
      'confirmationTokenExpiry',
      'firstName',
      'lastName',
      'nickname',
      'settings',
    ].forEach(value => {
      delete userClone[value];
    });


    map(userClone, (value, key) => {
      if (value || value === false) {
        if (key === 'email') {
          // Set Mixpanel special property $email and delete unused email property
          userClone.$email = user[key];
          delete userClone[key];
        } else if (value && key === 'gender') {
          // Change gender from integer to word equivalent ("male" / "female")
          userClone.gender = formattedGender;
        } else if (key === 'height') {
          // Change heightUnitPreference from integer to word equivalent ("IN" / "CM")
          userClone.heightUnitPreference = heightUnitPreference === height.units.IN ? 'IN' : 'CM';
        } else if (key === 'weight') {
          // Change heightUnitPreference from integer to word equivalent ("LB" / "KG")
          userClone.weightUnitPreference = weightUnitPreference === weight.units.LB ? 'LB' : 'KG';
        }
      } else {
        if (key === 'weight' || key === 'height') {
          // These unit preference properties will never be undefined
          // Remove them if their measurement values are undefined
          delete userClone[`${key}UnitPreference`];
        }

        // Remove properties with undefined values
        delete userClone[key];
      }
    });

    this.set(userClone);
  },

  /**
   * Tracks an event
   * @param  {String}  event  Event name
   */
  track(event) {
    Mixpanel.track(event);
  },

  /**
   * Tracks an event with properties
   * @param  {String}  event       Event name
   * @param  {Object}  properties  An object containing properties relevant to the event
   */
  trackWithProperties(event, properties) {
    Mixpanel.trackWithProperties(event, properties);
  },

  /**
   * Tracks an error event and its properties
   * @param  {Object}  properties  An error event property object which contains the error message,
   *                               stack trace, file path, bluetooth state properties, and anything
   *                               else that you feel would be relevant to the error event.
   */
  trackError(properties) {
    // Track Bluetooth status to identify potential BT-related issues
    BluetoothService.getState((error, { state }) => {
      Mixpanel.trackWithProperties('reactNativeError', {
        bluetoothState: state,
        ...properties,
      });
    });
  },

  /**
   * Registers super properties, which Mixpanel will include with every event
   * @param  {Object}  properties  An object containing properties to pass on every event
   */
  registerSuperProperties(properties) {
    Mixpanel.registerSuperProperties(properties);
  },
};
