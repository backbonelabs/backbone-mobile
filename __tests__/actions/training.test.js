import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { asyncActionMiddleware } from 'redux-async-action';
import trainingActions from '../../app/actions/training';
import Mixpanel from '../../app/utils/Mixpanel';

describe('Training Actions', () => {
  const initialState = {
    training: {
      selectedPlanIdx: 0,
      selectedLevelIdx: 0,
      selectedSessionIdx: 0,
      selectedStepIdx: 0,
    },
  };

  const store = configureStore([asyncActionMiddleware, thunk])(initialState);

  beforeEach(() => {
    store.clearActions();
    Mixpanel.trackWithProperties.mockClear();
  });

  test('creates an action to select a level', async () => {
    const {
      selectedPlanIdx,
      selectedLevelIdx,
      selectedSessionIdx,
      selectedStepIdx,
    } = initialState.training;

    await store.dispatch(trainingActions.selectLevel(selectedLevelIdx + 2));
    expect(store.getActions()).toMatchSnapshot();
    expect(Mixpanel.trackWithProperties).toHaveBeenCalledWith('selectLevel', {
      selectedPlanIdx,
      selectedLevelIdx,
      selectedSessionIdx,
      selectedStepIdx,
      newLevelIdx: selectedLevelIdx + 2,
    });
  });

  test('creates an action to select a session', async () => {
    const {
      selectedPlanIdx,
      selectedLevelIdx,
      selectedSessionIdx,
      selectedStepIdx,
    } = initialState.training;

    await store.dispatch(trainingActions.selectSession(selectedSessionIdx + 2));
    expect(store.getActions()).toMatchSnapshot();
    expect(Mixpanel.trackWithProperties).toHaveBeenCalledWith('selectSession', {
      selectedPlanIdx,
      selectedLevelIdx,
      selectedSessionIdx,
      selectedStepIdx,
      newSessionIdx: selectedSessionIdx + 2,
    });
  });
});
