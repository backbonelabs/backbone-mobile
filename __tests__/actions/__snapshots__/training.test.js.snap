// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Training Actions creates an action to select a level 1`] = `
Array [
  Object {
    "payload": 2,
    "type": "SELECT_LEVEL",
  },
]
`;

exports[`Training Actions creates an action to select a session 1`] = `
Array [
  Object {
    "payload": 2,
    "type": "SELECT_SESSION",
  },
]
`;
